### ISSUE TITLE
Setting confluence_max_doc_depth = 0 resolves index twice

### ISSUE THREAD

**Issue by jessetan** [sphinx-contrib/confluencebuilder#305]
**Problem**

If `confluence_max_doc_depth` is set to 0, all pages included using `toctree` are inlined into the index page. This causes the `doctree-resolved` event to be emitted twice.
This event is used by extensions to [modify the doctree in-place](https://www.sphinx-doc.org/en/master/extdev/appapi.html#event-doctree-resolved). If the event is emitted twice, any listeners will also be executed twice, which can have unwanted consequences.

`confluence_max_doc_depth` is set to 1, this does not happen.

**Reproduction**

index.rst:
```rst
Index page
==========

.. toctree::

   page
```

page.rst:
```rst
Page title
----------

Contents of page
```

conf.py:
```py
extensions = [
    'sphinxcontrib.confluencebuilder'
]
confluence_max_doc_depth = 0
```

To observe the event being emitted twice, change the following in __init__.py of confluencebuilder:
```py
def _doctree_resolved(app, doctree, docname):
    ConfluenceLogger.warn("Confluencebuilder doctree-resolved for document %s", docname)

def setup(app):
    ConfluenceLogger.initialize()
    app.connect('doctree-resolved', _doctree_resolved)

    # Rest of code continues as normal here
```

**Comment by jdknight** [comment]
Yup -- `doctree-resolved` is unfortunately being triggered more than once.

It appears to be an issue with supporting the `confluence_max_doc_depth` capability with asset management. When `confluence_max_doc_depth` is invoked and a doctree is modified, it will eventually trigger `env.resolve_references`; which in turn performs all post-transformations on a document (which will emit a signal). However, once all documents have been prepared and move to the writing stage, the write event (inside Sphinx) will then trigger all post-transformations (which emits the signal again).

This extension should not be triggering `env.resolve_references` in the builder's `process_tree_structure` call. I cannot recall why this was done, but with some crude testing, it may be related to the asset management support in this extension. Specially, it may have been added to mask an issue where this extension's doc-resolve hook is processed before other doc-resolve hooks. It appears that there are conditions which can prevent embedded/generated assets from being registered into the asset management before the translation phase is processed (see also #235). Resolving documents ahead of time (which in turn properly populates the asset management), prevents any missing asset issues in the translator -- which is possibly why this change was "lazily" added. A quick test shows that dropping the `env.resolve_references` call while also removing any embedded/generated image assets from the documentation set, all the unit tests appear to pass (although, a proper check would need to be made if there was another case for why a resolve call was being made).

At this time, will assume that the issue is with the asset management, which when resolved, the max-doc-depth feature can be adjusted to prevent unnecessary trigger events.

On a related note, with the introduction of `singleconfluence` builder in this extension (currently on the `master` branch), it may be ideal to deprecated the `confluence_max_doc_depth` option when it is set to a value of `0` (advising users to use the alternative builder instead).

Thanks for the report and the respective example to reproduce.

**Comment by jessetan** [comment]
Thanks @jdknight, I see that `singleconfluence` builder does not have the issue I reported. It would fit my needs just as well as setting `confluence_max_doc_depth` to `0`. Is it intended to have feature parity with the non-single builder? My first test showed at least #264 not working in `singleconfluence`

**Comment by jdknight** [comment]
The existing implementation (IMHO) for `confluence_max_doc_depth` is not ideal -- it does a series of unique things (doctree caching, temporarily overriding methods, etc.). Therefore, it would be probably best to promote the use of `singleconfluence` over the configuration scenario where `confluence_max_doc_depth = 0`. It is assumed that the capabilities provided by  `confluence_max_doc_depth` are still valid to some users -- specifically to squash large doctrees into a smaller set (but still desire some level of hierarchy).

There maybe a series of features that may not work with `singleconfluence`. The goal is to help weed out as many issues as possible before v1.3 is introduced (no timeline yet). At this time, the only issue noted (locally to myself) is limitations with `sphinx.ext.autosummary` and the builder (which will be a simple fix). Good to know that `confluence_add_secnumbers` (and related) capabilities are broken as well. Feel free to make an issue if desired  (and for any other oddity observed) -- in any case, will note to ensure the builder and that feature are explicitly tested (most likely with an enhanced unit test check).

**Comment by jessetan** [comment]
I've created a PR to fix the secnumbers #309

**Comment by jessetan** [comment]
I'm closing this bug, it does not make sense to spend time to fix this since the behavior will be superseded by the `singleconfluence` builder
